ui/src/app/formats.ts

```diff
@@ -0,0 +1,58 @@
+export interface Format {
+  id: string;
+  text: string;
+  qualities: Quality[];
+}
+
+export interface Quality {
+  id: string;
+  text: string;
+  value: string;
+  fmt?: string;
+}
+
+export const Formats: Format[] = [
+  {
+    id: 'any',
+    text: 'Any',
+    qualities: [],
+  },
+  {
+    id: 'mp4',
+    text: 'MP4',
+    qualities: [
+      { id: "1", value: 'best', text: 'Best MP4' },
+      { id: "2", value: '1440', text: '1440p' },
+      { id: "3", value: '1080', text: '1080p' },
+      { id: "4", value: '720', text: '720p' },
+      { id: "5", value: '480', text: '480p' },
+    ],
+  },
+  {
+    id: 'mp3',
+    text: 'MP3',
+    qualities: [
+      { id: "6", value: 'best', text: 'Best MP3' },
+      { id: "7", value: '128', text: '128 kbps' },
+      { id: "8", value: '192', text: '192 kbps' },
+      { id: "9", value: '320', text: '320 kbps' },
+    ],
+  },
+];
+
+export const fillQualities = (formats: Format[]): Format[] => {
+  let allQualities: Quality[] = [];
+  formats.forEach((fmt) => {
+    fmt.qualities = fmt.qualities.map((ql) => ({ ...ql, fmt: fmt.id }));
+    allQualities = allQualities.concat(fmt.qualities);
+  });
+
+  formats.find((format) => format.id === 'any').qualities = allQualities;
```

Comment:
Right now if "Any" format is selected, then the qualities contain both MP3 and MP4 qualities. This is not what we want.
What we want is when Any is selected, to have the same qualities (i.e. video resolutions) as when MP4 is selected. And it should be named "Best".
The audio one can also be named "Best" instead of "Best MP3", since it shows when "MP3" is in the format, so "Best" in the quality is enough..

Reply:
Sorry. I'm not sure I follow...
Shouldn't "Any" show all possible qualities, not just for a specific format?
I have is say "Best MP3" and "Best MP4" so the user knows which format it originates from.

It then passes down the original format as the called one, e.g. if format was "any" and text was "Best MP3" it would be the equivalent of selecting "MP3" as the format, and selecting "Best" as the quality.  

Are you saying that you want any to just show the same things as selecting "MP4"?

Or do you mean that "Any" should not target a specific file type, and would allow file types like "mkv" but with the same qualities available as "mp4"?

If it is the second option, would you prefer to leave quality options as are but not force a specific video format, or make it  only show video options?

Reply:
"Any", until now, meant "any format", while "MP4" forced MPEG4. So you would select the maximum resolution you want, and in the format you'd specify whether you want specifically MPEG4 or "Any" if it doesn't matter.
Now that we're adding MP3, we have to show audio qualities under quality, but I don't want the video options affected by that, so the above should keep working the same way as before.

There is a bit of a mixup here with regards to video vs. audio downloads. Perhaps we should think of a better way to organize all this. I'll think a bit about it. Let me know if you have ideas too. But we definitely want to still allow just selecting "best" and "any" and have it download whatever's available. We have to remember that it's possible to download from many sites beside YouTube here, and for other sites which don't have such quality/format availability, selecting "best" and leaving it to the site to determine what we'll be getting is the only working option.

Reply:
Ahh I see where I got mixed up...
I thought "Any" just meant any format and quality from the ones available

Implementing "any" like that would simplify everything significantly, as we don't need to add and find the format the quality came from.

I'll change the code to move the any option back to working like that.

I'll also see if I can think of a better way to organize

Reply:
> Implementing "any" like that would simplify everything significantly, as we don't need to add and find the format the quality came from.

To clarify, I mean implementing "any" the way where it ignores the format is easier.
The way I was doing was more complicated as it required merging all the formats.

Before I work on fixing this, would you prefer to keep a audio quality in "any" similar to video quality?
Or is it better if we removed it?

Thank you for clarifying and sorry for the earlier mix up!